cloud/jenkins/psmdb_operator_aks_latest.groovy

```diff
@@ -1,28 +1,85 @@
+        if (!"$PILLAR_VERSION" && !"$IMAGE_MONGOD") {
+            error("Either PILLAR_VERSION or IMAGE_MONGOD should be provided for release run!")
+        }
     }
 }
 
-void getImage(String IMAGE_NAME) {
-    versions_file = "source/e2e-tests/release_images"
-    IMAGE = """${sh(
-        returnStdout: true,
-        script: "cat ${versions_file} | egrep \"${IMAGE_NAME}=\" | cut -d = -f 2 | tr -d \'\"\' "
-    ).trim()}"""
-    if ("$IMAGE") {
-        return "$IMAGE"
-    }
-    else {
-        error("Empty image is returned for $IMAGE_NAME. Check PILLAR_VERSION or content of file with images")
+void getParam(String PARAM_NAME) {
+    PARAM = sh(script: "cat $release_versions | grep -i $PARAM_NAME= | cut -d = -f 2 | tr -d \'\"\'", , returnStdout: true).trim()
+
+    if ("$PARAM") {
+        return "$PARAM"
+    } else {
+        error("$PARAM_NAME not found in params file $release_versions")
     }
 }
 
 void prepareNode() {
+    echo "=========================[ Cloning the sources ]========================="
+    git branch: 'master', url: 'https://github.com/Percona-Lab/jenkins-pipelines'
+    sh """
+        # sudo is needed for better node recovery after compilation failure
+        # if building failed on compilation stage directory will have files owned by docker user
+        sudo sudo git config --global --add safe.directory '*'
+        sudo git reset --hard
+        sudo git clean -xdf
+        sudo rm -rf source
+        cloud/local/checkout $GIT_REPO $GIT_BRANCH
+    """
+
+    echo "=========================[ Assigning images for release test ]========================="
+    if ("$RELEASE_RUN" == "YES") {
+        if ("$IMAGE_OPERATOR") {
+            echo "IMAGE_OPERATOR=$IMAGE_OPERATOR (from job parameters)"
+        } else {
+            IMAGE_OPERATOR = getParam("IMAGE_OPERATOR")
+            echo "IMAGE_OPERATOR=$IMAGE_OPERATOR (from params file)"
+        }
+
+        if ("$IMAGE_MONGOD") {
+            echo "IMAGE_MONGOD=$IMAGE_MONGOD (from job parameters)"
+        } else {
+            IMAGE_MONGOD = getParam("IMAGE_MONGOD${PILLAR_VERSION}")
+            echo "IMAGE_MONGOD=$IMAGE_MONGOD (from params file)"
+        }
+
+        if ("$IMAGE_BACKUP") {
+            echo "IMAGE_BACKUP=$IMAGE_BACKUP (from job parameters)"
+        } else {
+            IMAGE_BACKUP  =getParam("IMAGE_BACKUP")
+            echo "IMAGE_BACKUP=$IMAGE_BACKUP (from params file)"
+        }
+
+        if ("$IMAGE_PMM_CLIENT") {
+            echo "IMAGE_PMM_CLIENT=$IMAGE_PMM_CLIENT (from job parameters)"
+        } else {
+            IMAGE_PMM_CLIENT = getParam("IMAGE_PMM_CLIENT")
+            echo "IMAGE_PMM_CLIENT=$IMAGE_PMM_CLIENT (from params file)"
+        }
+
+        if ("$IMAGE_PMM_SERVER") {
+            echo "IMAGE_PMM_SERVER=$IMAGE_PMM_SERVER (from job parameters)"
+        } else {
+            IMAGE_PMM_SERVER = getParam("IMAGE_PMM_SERVER")
+            echo "IMAGE_PMM_SERVER=$IMAGE_PMM_SERVER (from params file)"
+        }
```

Comment:
can we extract this repeating logic to a function?

Reply:
Although in Groovy, you can dereference variables like
def value = binding.getVariable(paramName)
however Jenkins pipeline is trying to call a method (`binding.getVariable(String)`), but it is being rejected by Jenkins' Script Security Sandbox. The Jenkins sandbox restricts the use of certain Groovy methods for security purposes, and in this case, the method `getVariable` is not permitted. If you have a better idea, I am more than welcome.

Reply:
instead of `PARAM=binding.getVariable(PARAM_NAME)` does this work?
```
def paramValue = "${params[paramName]}"
```